lightning/src/routing/scoring.rs

```diff
@@ -460,25 +458,466 @@ impl<T: Time> Readable for ChannelFailure<T> {
+			min_liquidity_offset_msat,
+			max_liquidity_offset_msat,
+			capacity_msat,
+			last_updated: &mut self.last_updated,
+			now: T::now(),
+			half_life,
+		}
+	}
+}
+
+impl<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> DirectedChannelLiquidity<L, T, U> {
+	/// Returns the success probability of routing the given HTLC `amount_msat` through the channel
+	/// in this direction.
+	fn success_probability(&self, amount_msat: u64) -> Probability {
+		let max_liquidity_msat = self.max_liquidity_msat();
+		let min_liquidity_msat = core::cmp::min(self.min_liquidity_msat(), max_liquidity_msat);
+		if amount_msat > max_liquidity_msat {
+			Probability::Zero
+		} else if amount_msat < min_liquidity_msat {
+			Probability::One
+		} else {
+			let numerator = max_liquidity_msat + 1 - amount_msat;
+			let denominator = max_liquidity_msat + 1 - min_liquidity_msat;
+			if numerator == denominator {
+				Probability::One
+			} else {
+				Probability::Ratio { numerator, denominator }
+			}
+		}
+	}
+
+	/// Returns the lower bound of the channel liquidity balance in this direction.
+	fn min_liquidity_msat(&self) -> u64 {
+		self.decayed_offset_msat(*self.min_liquidity_offset_msat)
+	}
+
+	/// Returns the upper bound of the channel liquidity balance in this direction.
+	fn max_liquidity_msat(&self) -> u64 {
+		self.capacity_msat
+			.checked_sub(self.decayed_offset_msat(*self.max_liquidity_offset_msat))
+			.unwrap_or(0)
+	}
+
+	fn decayed_offset_msat(&self, offset_msat: u64) -> u64 {
+		self.now.duration_since(*self.last_updated).as_secs()
+			.checked_div(self.half_life.as_secs())
+			.and_then(|decays| offset_msat.checked_shr(decays as u32))
+			.unwrap_or(0)
+	}
+}
+
+impl<L: DerefMut<Target = u64>, T: Time, U: DerefMut<Target = T>> DirectedChannelLiquidity<L, T, U> {
+	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat`.
+	fn failed_at_channel(&mut self, amount_msat: u64) {
+		if amount_msat < self.max_liquidity_msat() {
+			self.set_max_liquidity_msat(amount_msat);
+		}
+	}
+
+	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat` downstream.
+	fn failed_downstream(&mut self, amount_msat: u64) {
+		if amount_msat > self.min_liquidity_msat() {
+			self.set_min_liquidity_msat(amount_msat);
+		}
+	}
+
+	/// Adjusts the channel liquidity balance bounds when successfully routing `amount_msat`.
+	fn successful(&mut self, amount_msat: u64) {
+		let max_liquidity_msat = self.max_liquidity_msat().checked_sub(amount_msat).unwrap_or(0);
+		self.set_max_liquidity_msat(max_liquidity_msat);
+	}
+
+	/// Adjusts the lower bound of the channel liquidity balance in this direction.
+	fn set_min_liquidity_msat(&mut self, amount_msat: u64) {
+		*self.min_liquidity_offset_msat = amount_msat;
+		*self.max_liquidity_offset_msat = if amount_msat > self.max_liquidity_msat() {
+			0
+		} else {
+			self.decayed_offset_msat(*self.max_liquidity_offset_msat)
+		};
+		*self.last_updated = self.now;
```

Comment:
I think (for `feature = std`) `self.now` will stay at the `Instant` in time in which it was initially created, and never advance: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6e0de8e43e5b36080bb3942e81ae84cc

Reply:
Yup, the directed version is ephemeral. I'm using a `now` field to avoid any inconsistencies resulting from calling `T::now` in different methods (e.g., here, `max_liquidity_msat`, and `decayed_offset_msat`) when near a decay.